presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java

```diff
@@ -824,45 +822,46 @@ public HashComputationSet withHashComputation(Optional<HashComputation> hashComp
         }
     }
 
-    public static Optional<HashComputation> computeHash(Iterable<VariableReferenceExpression> fields)
+    private static Optional<HashComputation> computeHash(Iterable<VariableReferenceExpression> fields, FunctionManager functionManager)
     {
         requireNonNull(fields, "fields is null");
         List<VariableReferenceExpression> variables = ImmutableList.copyOf(fields);
         if (variables.isEmpty()) {
             return Optional.empty();
         }
-        return Optional.of(new HashComputation(fields));
+        return Optional.of(new HashComputation(fields, functionManager));
     }
 
-    public static Optional<Expression> getHashExpression(List<VariableReferenceExpression> variables)
+    public static Optional<RowExpression> getHashExpression(FunctionManager functionManager, List<VariableReferenceExpression> variables)
     {
         if (variables.isEmpty()) {
             return Optional.empty();
         }
 
-        Expression result = new GenericLiteral(StandardTypes.BIGINT, String.valueOf(INITIAL_HASH_VALUE));
+        RowExpression result = constant(INITIAL_HASH_VALUE, BIGINT);
         for (VariableReferenceExpression variable : variables) {
-            Expression hashField = new FunctionCall(
-                    QualifiedName.of(HASH_CODE),
-                    Optional.empty(),
-                    false,
-                    ImmutableList.of(new SymbolReference(variable.getName())));
-
-            hashField = new CoalesceExpression(hashField, new LongLiteral(String.valueOf(NULL_HASH_CODE)));
-
-            result = new FunctionCall(QualifiedName.of("combine_hash"), ImmutableList.of(result, hashField));
+            RowExpression hashField = call(functionManager, HASH_CODE, BIGINT, variable);
+            hashField = orNullHashCode(hashField);
+            result = call(functionManager, "combine_hash", BIGINT, result, hashField);
         }
         return Optional.of(result);
     }
 
+    private static RowExpression orNullHashCode(RowExpression expression)
+    {
+        return new SpecialFormExpression(SpecialFormExpression.Form.COALESCE, BIGINT, expression, constant(NULL_HASH_CODE, BIGINT));
```

Comment:
Maybe add a `checkArgument` to make sure that `expression` has the expected type.

Reply:
Actually, the check will be not necessary since the expression will always constructed to have `BIGINT` above.

Reply:
I am more worried we constructed `CallExpression` different than `FunctionSignature`. Probably we can add checks in `TypeValidator`.

Reply:
Currently all places calling this function will pass in row expression with correct type, but this is not guaranteed in the future when code are modified. It's safer to have a check.

Reply:
Adding check in `TypeValidator` could help cover more cases. The two are not conflicting with each other.